torch/csrc/autograd/function.h

```diff
@@ -383,6 +409,7 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
 
   edge_list next_edges_;
   PyObject* pyobj_ = nullptr; // weak reference
+  std::shared_ptr<Node> parent_node_ = nullptr;
```

Comment:
Why not store that in the metadata object we already have?
That way:
- We don't add an extra field to Node (we create a lot of them, so it uses quite a but of memory).
- You don't need custom code to be able to access it from python.
- The logic of printing the parent's stack can be done in the anomaly mode code directly (this is better as we might want to print the stack from other places than the engine and we don't want to have to replicate that code).

Reply:
I have tried several things about that, but I got a few problems:

- Assigning C++ node as the parent metadata would create circular dependency (between `anomaly_mode.h` and `function.h`)
- When I tried assigning parent metadata to `current_evaluating_node->pyobj()`, the PyObject pointer is actually a nullptr. Apparently the PyObject is not created when the node is created (from `python_cpp_function.cpp`, it is created when `functionToPyObject` is called).
- Using `functionToPyObject` function instead of `node->pyobj()` when assigning parent metadata in `function.cpp` makes the compiler complains about undefined `torch::autograd::functionToPyObject`, even though I have included `python_cpp_function.h` (I'm not sure why and how to solve this).

If you have a better idea, I'll be happy to do that.

Reply:
You definitely need to use `functionToPyObject`.
Also you need to make sure you use it only from within a `python_*` file (the python-less builds don't have it available). So from within `python_anomaly_mode` it should work fine.
What is the exact error you get about the undefined `functionToPyObject` ?

Reply:
I have found a way to put the parent in metadata in the new commit. This is done by introducing a new method `AnomalyMetadata::assign_parent` which should take a shared pointer of Node. In PyAnomalyMetadata, the node is converted to PyObject and then stored in `metadata["parent_"]`. Please take a look.